sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,550 @@
+ * isprint() available so this assumes ASCII and drops the control characters
+ * that could could issues with inappropriate terminal settings at the host.
+ */
+static const uint32_t kCtrlChars =
+    ((uint32_t)1u << '\t') | ((uint32_t)1u << '\n') | ((uint32_t)1u << '\r');
+static const uint32_t kMap[0x100u / 0x20u] = {
+    // suppress control characters and ASCII 127 (Delete)
+    kCtrlChars, ~0u, ~0u, 0x7fffffffu,
+    // top-bit set characters are harmless; propagate them unchanged.
+    ~0u, ~0u, ~0u, ~0u};
+
+// Our stdout function
+static size_t usb_log(void *data, const char *buf, size_t len);
+
+/**
+ * Configuration values for USB.
+ */
+static uint8_t config_descriptors[kCfgDscrLenMax];
+
+// Since isprint() is unavailable, use a simple bitmap to test whether it is
+// safe to transmit the given ASCII character unmodified.
+static bool isprintable(uint8_t ch) {
+  return ((kMap[ch >> 5] >> (ch & 0x1fu)) & 1u) != 0u;
+}
+
+// Place all buffers into the free list
+static void free_list_init(usb_logging_ctx_t *ctx) {
+  usb_logging_buf_t *prev = NULL;  // End of list.
+  unsigned idx = kUSBLogNumBuffers;
+  // Order them such that the first buffer is initially at the head.
+  while (idx-- > 0u) {
+    ctx->buf[idx].next = prev;
+    prev = &ctx->buf[idx];
+  }
+  ctx->free = ctx->buf;
+}
+
+// Attempt to claim a buffer from the free list.
+static inline usb_logging_buf_t *buffer_claim(usb_logging_ctx_t *ctx) {
+  usb_logging_buf_t *buf = ctx->free;
+  if (buf) {
+    ctx->free = buf->next;
+    buf->next = NULL;
+    buf->bytes_used = 0u;
+  }
+  return buf;
+}
+
+// Return a buffer to the free list.
+static inline void buffer_release(usb_logging_ctx_t *ctx,
+                                  usb_logging_buf_t *buf) {
+  buf->next = ctx->free;
+  ctx->free = buf;
+}
+
+static status_t buffer_send(usb_logging_ctx_t *ctx, usb_logging_stream_t *s) {
+  TRY_CHECK(ctx && s && s->send_buf);
+
+  const uint8_t *data = s->send_buf->data;
+  uint16_t len = s->send_buf->bytes_used;
+
+  if (kDevLogging) {
+    LOG_INFO("sending %p %p %u\n", ctx, data, len);
+  }
+
+  TRY(usb_testutils_transfer_send(ctx->usbutils, s->ep, data, len,
+                                  kUsbTestutilsXfrDoubleBuffered));
+  // Have at least started to send the buffer...
+  s->sending = true;
+  return OK_STATUS();
+}
+
+static status_t buffer_completed(usb_logging_ctx_t *ctx,
+                                 usb_logging_stream_t *s) {
+  TRY_CHECK(s->wr_buf);
+
+  // Make the buffer available for sending.
+  usb_logging_buf_t **prev = &s->send_buf;
+  while (*prev) {
+    prev = &(*prev)->next;
+  }
+  *prev = s->wr_buf;
+  s->wr_buf = NULL;
+
+  // Attempt to send.
+  status_t res = OK_STATUS();
+  if (!s->sending) {
+    res = buffer_send(ctx, s);
+  }
+  return res;
+}
+
+// Indicates whether a stream is still busy, when attempting to finalize.
+static inline bool stream_busy(usb_logging_ctx_t *ctx,
```

Comment:
The name looks like it's just a status check, but it actually does a flushing + wait operation. This probably needs a better name, e.g. `poll_flush`?